[ARM] Save floating point registers and status registers with save_fp function attribute

pestctrl opened the pull request:
The interrupt attribute currently doesn't save floating-point related registers in the frame setup and destroy. 

This patch adds a new interrupt_save_fp attribute that will save floating-point registers as well.

pestctrl commented:
ping!

efriedma-quic commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -2239,6 +2239,20 @@ The semantics are as follows:
+Clang supports the GNU style ``__attribute__((interrupt_save_fp("TYPE")))``
+on ARM targets. This attribute behaves the same way as the ARM interrupt
+attribute, except the general purpose floating point registers are also saved. 
+If the FPEXC or FPSCR are needed, that state must be saved manually. Note, even 
```
This says FPSCR isn't saved... but some of the tests show FPSCR getting saved.  Which is right?

pestctrl commented:
ping!

efriedma-quic commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -338,10 +338,16 @@ def warn_anyx86_excessive_regsave : Warning<
+   "`interrupt_save_fp` only applies to targets that have a VFP unit enabled "
+   "for this compilation; this will be treated as a regular `interrupt` "
+   "attribute">,
+   InGroup<Extra>;
```
Please don't add diagnostics to Wextra; it's a specific set of diagnostics based on the equivalent gcc flag, not a catchall.

pestctrl replied:
Done

efriedma-quic commented on clang/test/CodeGen/arm-interrupt-save-fp-attr-status-regs.c:
```diff
@@ -0,0 +1,34 @@
+__attribute__((interrupt_save_fp)) void test_generic_interrupt() {
+    // CHECK-R:      vmrs	r4, fpscr
+    // CHECK-R-NEXT: vmrs	r5, fpexc
+    // CHECK-R-NEXT: .save  {fpscr, fpexc}
```
It looks like this is the only test that checks the new `.save` syntax? There should be tests in llvm/test/CodeGen/ARM and llvm/test/MC/ARM

pestctrl replied:
I have reverted that syntax change. We had initially rolled it out internally, and later found that our assembler didn't like that syntax change, so we rolled it back. I forgot to roll it back on the upstream branch. 

To our understanding, the .save directive is for unwinding, and only accepts GPR's. .vsave isn't the correct directive to use either, since those only accept DPR's. In the end, we decided to just have those .save's refer to the actual GPR's.

efriedma-quic replied:
At that point, you might as well not emit the .save at all?  At best, the debugger will just see the values get overwritten, and ignore the second .save.

efriedma-quic commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5584,6 +5584,20 @@ static void handleAbiTagAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+                                         const ParsedAttr &AL) {
+  handleARMInterruptAttr(S, D, AL);
+
+  bool VFP = S.Context.getTargetInfo().hasFeature("vfp");
```
Do we care if someone uses a "target" attribute?  Maybe not relevant for embedded targets where you'd use this.

efriedma-quic commented on llvm/lib/Target/ARM/ARMBaseRegisterInfo.cpp:
```diff
@@ -80,13 +80,47 @@ ARMBaseRegisterInfo::getCalleeSavedRegs(const MachineFunction *MF) const {
+    // float. This will prevent the Thumb1FrameLowering (cortex-m0) from
+    // crashing due to an llvm_unreachable being triggered when a D-class
+    // register is in the calling convention.
+    if (STI.isTargetHardFloat() && F.hasFnAttribute("save-fp")) {
```
It's a little weird to check isTargetHardFloat here... that just refers to whether the calling convention passes arguments in integer registers or FP registers.  It doesn't have anything to do with whether floating-point registers are used.

pestctrl replied:
How about we check for if if the subtarget has FP registers? 

```
if (STI.hasFPRegs() && F.hasFnAttribute("save-fp"))
```

efriedma-quic reviewed (approved):
LGTM

Commit:
[ARM][Thumb] Save FPSCR + FPEXC for save-vfp attribute

FPSCR and FPEXC will be stored in FPStatusRegs, after GPRCS2 has been
saved.

- GPRCS1
- GPRCS2
- FPStatusRegs (new)
- DPRCS
- GPRCS3
- DPRCS2

FPSCR is present on all targets with a VFP, but the FPEXC register is
not present on Cortex-M devices, so different amounts of bytes are
being pushed onto the stack depending on our target, which would
affect alignment for subsequent saves.

DPRCS1 will sum up all previous bytes that were saved, and will emit
extra instructions to ensure that its alignment is correct. My
assumption is that if DPRCS1 is able to correct its alignment to be
correct, then all subsequent saves will also have correct alignment.

Avoid annotating the saving of FPSCR and FPEXC for functions marked
with the interrupt_save_fp attribute, even though this is done as part
of frame setup.  Since these are status registers, there really is no
viable way of annotating this. Since these aren't GPRs or DPRs, they
can't be used with .save or .vsave directives. Instead, just record
that the intermediate registers r4 and r5 are saved to the stack
again.

Co-authored-by: Jake Vossen <jake@vossen.dev>
Co-authored-by: Alan Phipps <a-phipps@ti.com>

efriedma-quic commented:
Is this still waiting for something from me?

pestctrl commented:
@efriedma-quic No, I will push this on Monday. Thank you!
